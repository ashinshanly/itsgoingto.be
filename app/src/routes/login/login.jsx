import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { browserHistory } from 'react-router'
import { withCookies, Cookies } from 'react-cookie'
import { isEmpty, dissoc, join } from 'ramda'
import { postLogin, APIError } from 'store/api'
import { hasUserSelector } from 'store/user'
import { setLoading } from 'store/loader'
import Button from 'components/button'
import EventBus from 'components/event-bus'
import Back from 'components/back'
import './login.scss'

const KEY_ENTER = 13

class Login extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      data : {
        username : '',
        password : '',
      },
      errors : { },
      loading: false,
    }

    if (props.hasUser) {
      props.setLoading(true)
      this.state.loading = true
      browserHistory.push('/admin')
    }
  }

  componentWillMount = () => {
    const { setLoading } = this.props

    setLoading(false)
    this.eventBus = EventBus.getEventBus()
  }

  handleChange = ({ target }) => {
    const data = { ...this.state.data, [target.id] : target.value }

    this.setState({ data })
    this.validate(true, data)
  }

  validate = (removeOnly = false, data) => {
    if (!data) {
      data = this.state.data
    }
    const { username, password } = data
    let { errors } = this.state

    errors = dissoc('api', errors)

    if (username.length === 0) {
      if (!removeOnly) {
        errors.username = 'Please enter a username.'
      }
    } else {
      errors = dissoc('username', errors)
    }

    if (password.length === 0) {
      if (!removeOnly) {
        errors.password = 'Please enter a password.'
      }
    } else {
      errors = dissoc('password', errors)
    }

    this.setState({ errors })

    return isEmpty(errors)
  }

  handleKeyPress = (event) => {
    event = event || window.event
    const key = event.keyCode || event.charCode

    if (key === KEY_ENTER) {
      this.eventBus.emit('login-submit')
    }
  }

  submit = () => {
    const { data : { username, password } } = this.state
    const { postLogin, cookies, setLoading } = this.props

    if (this.validate()) {
      return postLogin(username, password).then((response) => {
        if (response instanceof APIError) {
          const errors = {}

          try {
            errors.api = join('<br />', response.details.error.errors)
          } catch (err) {
            errors.api = 'There was an unknown error.'
          }
          this.setState({ errors })
          return true
        } else {
          setLoading(true)
          cookies.set('itsgoingtobeUserToken', response.token, { path: '/', maxAge: 3600 })
          browserHistory.push('/admin')
          return false
        }
      })
    } else {
      return Promise.resolve()
    }
  }

  render () {
    const { data : { username, password }, errors, loading } = this.state

    if (loading) {
      return <div />
    }

    join('-', ['1', '2', '3'])

    return (
      <div>
        <Back />
        <div className='container header-container'>
          <div className='header center-text'>
            <h1>Login</h1>
          </div>
        </div>
        <div className='container login-container'>
          <div className={'input input-username' + (errors.username || errors.api ? ' input-error' : '')}>
            <label className='input-label input-label-username' htmlFor='question'>Username</label>
            <input
              className='input-field input-field-username'
              type='text'
              id='username'
              name='username'
              ref='username'
              value={username}
              onChange={this.handleChange} />
            <span className='input-error-label'>{errors.username}</span>
          </div>
          <div className={'input input-password' + (errors.username || errors.api ? ' input-error' : '')}>
            <label className='input-label input-label-password' htmlFor='question'>Password</label>
            <input
              className='input-field input-field-password'
              type='password'
              id='password'
              name='password'
              ref='password'
              value={password}
              onKeyDown={this.handleKeyPress}
              onChange={this.handleChange} />
            <span className='input-error-label'>{errors.password || errors.api}</span>
          </div>
          <Button
            className='pull-right btn--small'
            text='Login'
            disabled={!isEmpty(errors)}
            callback={this.submit}
            submitEvent='login-submit' />
        </div>
      </div>
    )
  }
}

Login.propTypes = {
  postLogin  : PropTypes.func.isRequired,
  hasUser    : PropTypes.bool.isRequired,
  setLoading : PropTypes.func.isRequired,
  cookies    : PropTypes.instanceOf(Cookies).isRequired,
}

const mapStateToProps = (state) => ({
  hasUser : hasUserSelector(state),
})

const mapDispatchToProps = (dispatch) => ({
  postLogin  : (username, password) => dispatch(postLogin(username, password)),
  setLoading : (value) => dispatch(setLoading(value)),
})

export default connect(mapStateToProps, mapDispatchToProps)(withCookies(Login))
