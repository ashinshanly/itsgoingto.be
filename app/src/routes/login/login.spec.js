/* global expect, jest */
import React from 'react'
import { shallow } from 'enzyme'
import { Cookies } from 'react-cookie'
import { Login } from './login'

const props = {
  postLogin: jest.fn(),
  hasUser: false,
  setLoading: jest.fn(),
  clearUser: jest.fn(),
  cookies: new Cookies(),
}

describe('(Route) Login', () => {
  // describe('(Lifecycle) componentDidMount', () => {
  //   describe('has user', () => {
  //     it('should call clearUser', () => {
  //     })

  //     it('should remove cookie', () => {
  //     })
  //   })
  //   it('should set loading false', () => {
  //   })
  // })

  // describe('(Action) Submit click', () => {
  //   describe('validation', () => {
  //     it('should resolve empty promise if validations fails', () => {
  //     })

  //     describe('username', () => {
  //       it('should add error if fails validation', () => {
  //       })

  //       it('should remove error if fails validation', () => {
  //       })
  //     })

  //     describe('password', () => {
  //       it('should add error if fails validation', () => {
  //       })

  //       it('should remove error if fails validation', () => {
  //       })
  //     })
  //   })

  //   it('should call postLogin if validation passes', () => {
  //   })

  //   describe('postLogin success', () => {
  //     it('should call setLoading with true', () => {
  //     })

  //     it('should set a user token cookie', () => {
  //     })

  //     it('should redirect to /admin', () => {
  //     })
  //   })

  //   describe('postLogin failure', () => {
  //     it('should set state with api errors', () => {
  //     })

  //     it('should set state with unknown error', () => {
  //     })
  //   })
  // })

  // describe('(Action) Username onChange', () => {
  //   it('should update state', () => {
  //   })

  //   it('should remove error if passes validation', () => {
  //   })

  //   it('should not remove error if fails validation', () => {
  //   })
  // })

  // describe('(Action) Password onChange', () => {
  //   it('should update state', () => {
  //   })

  //   it('should remove error if passes validation', () => {
  //   })

  //   it('should not remove error if fails validation', () => {
  //   })
  // })

  // describe('(Action) Password onKeyPress', () => {
  //   it('should emit \'login-submit\' to eventBus', () => {
  //   })
  // })

  describe('(Render)', () => {
    describe('initial login page', () => {
      it('matches snapshot', () => {
        const wrapper = shallow(<Login {...props} />)

        expect(wrapper).toMatchSnapshot()
      })
    })

    // describe('with data', () => {
    //   it('matches snapshot', () => {
    //     const wrapper = shallow(<Login {...props} />)

    //     expect(wrapper).toMatchSnapshot()
    //   })
    // })

    // describe('with username error', () => {
    //   it('matches snapshot', () => {
    //     const wrapper = shallow(<Login {...props} />)

    //     expect(wrapper).toMatchSnapshot()
    //   })
    // })

    // describe('with password error', () => {
    //   it('matches snapshot', () => {
    //     const wrapper = shallow(<Login {...props} />)

    //     expect(wrapper).toMatchSnapshot()
    //   })
    // })

    // describe('with api error', () => {
    //   it('matches snapshot', () => {
    //     const wrapper = shallow(<Login {...props} />)

    //     expect(wrapper).toMatchSnapshot()
    //   })
    // })

    // describe('with loading', () => {
    //   it('matches snapshot', () => {
    //     const wrapper = shallow(<Login {...props} />)

    //     expect(wrapper).toMatchSnapshot()
    //   })
    // })
  })
})
