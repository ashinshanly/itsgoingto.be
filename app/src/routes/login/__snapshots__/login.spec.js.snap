// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Route) Login (Render) initial login page matches snapshot 1`] = `
<div>
  <Back />
  <div
    className="container header-container"
  >
    <div
      className="header center-text"
    >
      <h1>
        Login
      </h1>
    </div>
  </div>
  <div
    className="container login-container"
  >
    <div
      className="input input-username"
    >
      <label
        className="input-label input-label-username"
        htmlFor="question"
      >
        Username
      </label>
      <input
        className="input-field input-field-username"
        id="username"
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
      <span
        className="input-error-label"
      />
    </div>
    <div
      className="input input-password"
    >
      <label
        className="input-label input-label-password"
        htmlFor="question"
      >
        Password
      </label>
      <input
        className="input-field input-field-password"
        id="password"
        name="password"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="password"
        value=""
      />
      <span
        className="input-error-label"
      />
    </div>
    <Button
      callback={[Function]}
      className="pull-right btn--small"
      disabled={false}
      submitEvent="login-submit"
      text="Login"
    />
  </div>
</div>
`;
